src/middleware/middleware.rs

```diff
@@ -0,0 +1,23 @@
+use std::{future::Future, pin::Pin, sync::Arc};
+
+use super::Next;
+use crate::{response::IntoResponse, Request, Response, Result};
+
+pub(crate) type ArcMiddleware<State> = Arc<dyn Middleware<State>>;
+pub type BoxFuture<T> = Pin<Box<dyn Future<Output = T> + Send + 'static>>;
+
+pub trait Middleware<State>: Send + Sync + 'static {
+    fn call(&self, request: Request<State>, next: Next<State>) -> BoxFuture<Result<Response>>;
+}
+
+impl<State, T, F> Middleware<State> for T
+where
+    T: Fn(Request<State>, Next<State>) -> F + Send + Sync + 'static,
```

Comment:
I'm contemplating putting the bound `Copy` on `T` to prevent middleware functions from capturing their environment. Considering how complex the route definitions can become for an application, it seems like a good idea to require users to write a `struct` with a `Middleware` impl so the intent is clear.